service/history/execution/mutable_state_task_generator_test.go

```diff
@@ -667,6 +771,931 @@ func (s *mutableStateTaskGeneratorSuite) TestGetNextDecisionTimeout() {
+		{
+			name: "Success case - scheduleToStartTimeout not 0",
+			setupMock: func() {
+				s.mockMutableState.EXPECT().GetDecisionInfo(decisionScheduleID).Return(decision, true).Times(1)
+				s.mockMutableState.EXPECT().AddTransferTasks(&persistence.DecisionTask{
+					TaskData: persistence.TaskData{
+						Version: decision.Version,
+					},
+					DomainID:   executionInfo.DomainID,
+					TaskList:   decision.TaskList,
+					ScheduleID: decision.ScheduleID,
+				}).Times(1)
+				scheduleToStartTimeout := time.Duration(1)
+				s.mockMutableState.EXPECT().GetDecisionScheduleToStartTimeout().Return(scheduleToStartTimeout).Times(1)
+				s.mockMutableState.EXPECT().AddTimerTasks(&persistence.DecisionTimeoutTask{
+					TaskData: persistence.TaskData{
+						VisibilityTimestamp: time.Unix(0, decision.ScheduledTimestamp).Add(scheduleToStartTimeout),
+						Version:             decision.Version,
+					},
+					TimeoutType:     int(TimerTypeScheduleToStart),
+					EventID:         decision.ScheduleID,
+					ScheduleAttempt: decision.Attempt,
+				}).Times(1)
+			},
+		},
+		{
+			name: "Error case - GetDecisionInfo error",
+			setupMock: func() {
+				s.mockMutableState.EXPECT().GetDecisionInfo(decisionScheduleID).Return(nil, false).Times(1)
+			},
+			err: &types.InternalServiceError{
+				Message: fmt.Sprintf("it could be a bug, cannot get pending decision: %v", decisionScheduleID),
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		s.T().Run(tc.name, func(t *testing.T) {
+			s.mockMutableState.EXPECT().GetExecutionInfo().Return(executionInfo).Times(1)
+			tc.setupMock()
+
+			err := s.taskGenerator.GenerateDecisionScheduleTasks(decisionScheduleID)
+
+			if tc.err != nil {
+				s.Error(err)
+				s.Equal(tc.err, err)
+			} else {
+				s.NoError(err)
+			}
+		})
+
+	}
+}
+
+func (s *mutableStateTaskGeneratorSuite) TestGenerateDecisionStartTasks() {
+	seed := int64(1)
+	rand.Seed(seed)
+	decisionScheduleID := int64(123)
+	decision := &DecisionInfo{
+		Version:          constants.TestVersion,
+		ScheduleID:       123,
+		StartedTimestamp: time.Now().UnixNano(),
+	}
+
+	testCases := []struct {
+		name      string
+		setupMock func()
+		err       error
+	}{
+		{
+			name: "Success case - attempt less or equal to 1",
+			setupMock: func() {
+				decision.Attempt = 0
+				decision.DecisionTimeout = 1
+				s.mockMutableState.EXPECT().GetDecisionInfo(decisionScheduleID).Return(decision, true).Times(1)
+				s.mockMutableState.EXPECT().AddTimerTasks(&persistence.DecisionTimeoutTask{
+					TaskData: persistence.TaskData{
+						VisibilityTimestamp: time.Unix(0, decision.StartedTimestamp).Add(time.Duration(decision.DecisionTimeout) * time.Second),
+						Version:             decision.Version,
+					},
+					TimeoutType:     int(TimerTypeStartToClose),
+					EventID:         decision.ScheduleID,
+					ScheduleAttempt: decision.Attempt,
+				})
+			},
+		},
+		{
+			name: "Success case - attempt greater than 1",
+			setupMock: func() {
+				decision.Attempt = 2
```

Comment:
this might make the next test a bit wierd because it's mutating a decision that's shared and it'll be 2 for other subsequent tests (and then now the test ordering matters - they probably can't be run in any order).  Any chance you can return a new struct?